services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/crossregion/S3CrossRegionSyncClientRedirectTest.java

```diff
@@ -0,0 +1,269 @@
+        assertThat(listObjectsResponse.contents()).isEqualTo(S3_OBJECTS);
+
+        ArgumentCaptor<ListObjectsRequest> requestArgumentCaptor = ArgumentCaptor.forClass(ListObjectsRequest.class);
+        verify(mockDelegateClient, times(2)).listObjects(requestArgumentCaptor.capture());
+        assertThat(requestArgumentCaptor.getAllValues().get(0).overrideConfiguration()).isNotPresent();
+
+        EndpointProvider overridenEndpointProvider =
+            requestArgumentCaptor.getAllValues().get(1).overrideConfiguration().get().endpointProvider().get();
+        assertThat(overridenEndpointProvider).isInstanceOf(BucketEndpointProvider.class);
+
+        assertThat(((S3EndpointProvider) overridenEndpointProvider)
+                       .resolveEndpoint(e -> e.region(Region.US_WEST_2).bucket(CROSS_REGION_BUCKET).build()).get().url().getHost())
+            .isEqualTo("s3." + CROSS_REGION + ".amazonaws.com");
+    }
+
+    @Test
+    void decoratorUsesCache_IfCrossRegionAlreadyPresent() throws Exception {
+        when(mockDelegateClient.serviceClientConfiguration()).thenReturn(ENDPOINT_CONFIGURED);
+        when(mockDelegateClient.listObjects(any(ListObjectsRequest.class)))
+            .thenThrow(redirectException(301, CROSS_REGION))
+            .thenReturn(ListObjectsResponse.builder().contents(S3_OBJECTS).build())
+            .thenReturn(ListObjectsResponse.builder().contents(S3_OBJECTS).build());
+
+        // first call to update the local cache
+        ListObjectsResponse preCacheCall = decoratedS3Client.listObjects(i -> i.bucket(CROSS_REGION_BUCKET));
+        assertThat(preCacheCall.contents()).isEqualTo(S3_OBJECTS);
+        // Second call to use the local cache
+        ListObjectsResponse postCacheCall = decoratedS3Client.listObjects(i -> i.bucket(CROSS_REGION_BUCKET));
+        assertThat(postCacheCall.contents()).isEqualTo(S3_OBJECTS);
+
+        ArgumentCaptor<ListObjectsRequest> requestArgumentCaptor = ArgumentCaptor.forClass(ListObjectsRequest.class);
+        verify(mockDelegateClient, times(3)).listObjects(requestArgumentCaptor.capture());
+        assertThat(requestArgumentCaptor.getAllValues().get(0).overrideConfiguration()).isNotPresent();
+
+        EndpointProvider overridenEndpointProviderPreCache =
+            requestArgumentCaptor.getAllValues().get(1).overrideConfiguration().get().endpointProvider().get();
+        assertThat(overridenEndpointProviderPreCache).isInstanceOf(BucketEndpointProvider.class);
+
+        assertThat(((S3EndpointProvider) overridenEndpointProviderPreCache)
+                       .resolveEndpoint(e -> e.region(Region.US_WEST_2).bucket(CROSS_REGION_BUCKET).build()).get().url().getHost())
+            .isEqualTo("s3." + CROSS_REGION + ".amazonaws.com");
+
+        EndpointProvider overriddenEndpointProviderPostCache =
+            requestArgumentCaptor.getAllValues().get(1).overrideConfiguration().get().endpointProvider().get();
```

Comment:
This call to captor gets the same number `1` as above. 

In general, I am not sure which of these assertions that really show the test case behavior, that the cache is used when the second call is made? I think grouping or removing some of the nice-to-have assertions together will make that easier to see.